src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,299 @@
+import org.xbill.DNS.Section;
+import org.xbill.DNS.SetResponse;
+import org.xbill.DNS.Type;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+@Builder(buildMethodName = "innerBuild")
+public class LookupSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+  private static final int DEFAULT_NDOTS = 1;
+  /** The {@link Resolver} to use to look up records. */
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}.
+   */
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@code 1}, meaning meaning that if there are any dots in a name, the
+   * name will be tried first as an absolute name.
+   */
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  /** Configures the search path used to look up relative names with less than ndots dots. */
+  @Singular("searchPathElement")
+  private final List<Name> searchPath;
+
+  /**
+   * If set to true, cached results with multiple records will be returned with the starting point
+   * shifted one step per request.
+   */
+  @Builder.Default private final boolean cycleResults = false;
+
+  /**
+   * Configures the Cache instances to be used for lookups for the different {@link DClass} values.
+   */
+  @Singular("cache")
+  private final Map<Integer, Cache> caches;
+
+  // this augments the lombok generated code.
+  public static class LookupSessionBuilder {
+    public LookupSession build() {
+      // note that this transform is idempotent, as concatenating an already absolute Name with root
+      // is a noop.
+      if (searchPath != null) {
+        this.searchPath =
+            searchPath.stream()
+                .map(
+                    name -> {
+                      try {
+                        return Name.concatenate(name, Name.root);
+                      } catch (NameTooLongException e) {
+                        throw new IllegalArgumentException("Search path name too long");
+                      }
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
+      }
+      return innerBuild();
+    }
+  }
+
+  /**
+   * Make an asynchronous lookup of the provided name.
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in {@link Type}.
+   * @param dclass the class to look up, values should correspond to constants in {@link DClass}.
+   * @return A {@link CompletionStage} what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
```

Comment:
add an overload without DClass that defaults to DClass.IN, and maybe one that takes a String instead of a Name